tratto/src/main/java/star/tratto/util/javaparser/JavaParserUtils.java

```diff
@@ -0,0 +1,1078 @@
+            );
+            logger.error(errMsg);
+            throw new JPClassNotFoundException(errMsg);
+        }
+    }
+
+    /**
+     * Gets a compilation unit from a Java file path.
+     *
+     * @param path a Java file
+     * @return the corresponding JavaParser compilation unit. Returns
+     * {@code Optional.empty()} if an error occurs while attempting to parse
+     * the file.
+     */
+    public static Optional<CompilationUnit> getCompilationUnit(Path path) {
+        try {
+            return javaParser.parse(path).getResult();
+        } catch (IOException e) {
+            return Optional.empty();
+        }
+    }
+
+    /**
+     * Checks if a method is non-private and non-static.
+     *
+     * @param methodUsage a method
+     * @return true iff the method is non-private non-static
+     */
+    private static boolean isNonPrivateNonVoidMethod(MethodUsage methodUsage) {
+        return !methodUsage.getDeclaration().getReturnType().isVoid() && !methodUsage.getDeclaration().accessSpecifier().equals(AccessSpecifier.PRIVATE);
+    }
+
+    /**
+     * Checks if an attribute is non-private and non-static. An attribute
+     * is any class field.
+     *
+     * @param fieldDeclaration a class field
+     * @return true iff the field is non-private and non-static
+     */
+    public static boolean isNonPrivateNonStaticAttribute(ResolvedFieldDeclaration fieldDeclaration) {
+        return !fieldDeclaration.accessSpecifier().equals(AccessSpecifier.PRIVATE) && !fieldDeclaration.isStatic();
+    }
+
+    public static boolean isNonPrivateStaticNonVoidMethod(MethodUsage methodUsage) {
+        return methodUsage.getDeclaration().isStatic() && isNonPrivateNonVoidMethod(methodUsage);
+    }
+
+    public static boolean isNonPrivateNonStaticNonVoidMethod(MethodUsage methodUsage) {
+        return !methodUsage.getDeclaration().isStatic() && isNonPrivateNonVoidMethod(methodUsage);
+    }
+
+    /**
+     * Unfortunately, JavaParser does not allow to parse constructors as method declarations. Since we don't
+     * need the constructor declaration for anything, if the passed methodSourceCode is from a constructor,
+     * this method returns null. This behavior must be handled by the caller.
+     * @return null if the passed method source code is actually a constructor
+     * @throws IllegalArgumentException if the provided methodSourceCode cannot be parsed by JavaParser
+     */
+    public static MethodDeclaration getMethodDeclaration(String methodSourceCode) throws IllegalArgumentException {
+        try {
+            return javaParser.parseBodyDeclaration(methodSourceCode).getResult().get().asMethodDeclaration();
+        } catch (NoSuchElementException e) {
+            throw new IllegalArgumentException("JavaParser cannot parse:" + System.lineSeparator() + methodSourceCode, e);
+        } catch (IllegalStateException e) {
+            return null; // This happens when the methodSourceCode is actually a constructor
```

Comment:
Are there any other situations, such as when the code is non-syntactic?
Please don't use exceptions for control flow when there is a simple boolean test that you can do to check the situation.

Reply:
@AML14